From 7185eeb05afb095b30a2b239525734d380e9f4b5 Mon Sep 17 00:00:00 2001
From: Luca Stefani <luca.stefani.ge1@gmail.com>
Date: Tue, 6 Nov 2018 15:26:06 +0100
Subject: [PATCH 3/3] [SQUSH][DNM] Merge tag 'android-9.0.0_r16' into
 lineage-16.0
MIME-Version: 1.0
Content-Type: text/plain; charset=UTF-8
Content-Transfer-Encoding: 8bit

commit 9576e6f5046d4c310f9fae51fdbb6571928a271b
Merge: 892883806 ad80f594d
Author: android-build-team Robot <android-build-team-robot@google.com>
Date:   Wed Sep 5 23:35:47 2018 +0000

    Merge cherrypicks of [4948448, 4948511, 4948299, 4947854, 4947306, 4946150, 4948531, 4946116, 4948532] into pi-qpr1-release

    Change-Id: I0da058074593bf706f43d8d6389a8e30bb46b09a

commit ad80f594d4dec7d9ef4775abbdde12d53a5876f4
Author: Tyler Gunn <tgunn@google.com>
Date:   Tue Aug 28 08:40:15 2018 -0700

    Handle carrier config changes in TelecomAccountRegistry.

    TelecomAccountRegistry makes use of the carrier configuration to set a
    number of properties on the Telephony phone account(s).  This is done when
    there is a subscription change.
    It is, however, possible for a carrier config change to occur outside of
    a subscription change.  If this happens, the change will not be reflected
    in the accounts registered by the TelecomAccountRegistry.
    Added a receiver for ACTION_CARRIER_CONFIG_CHANGED which triggers a
    rebuild of the telephony phone account.  To ensure that unnecessary updates
    are not made to the phone account in Telecom, compare the new to the old
    phone account definition in this case.  If there is no change, we do not
    update the phone account.

    Test: Manual testing -- verify that carrier config change causes update to
    the phone account for Telephony.
    Bug: 112884689
    Merged-In: I8c7a1ddee811980c4e75815aa6a55c9df11395a6
    Change-Id: I8c7a1ddee811980c4e75815aa6a55c9df11395a6
    (cherry picked from commit 0fceba4a4c8c9c6ea828e4a8d1da95e32b3280ad)
    (cherry picked from commit 15daf606bf0b2f018df7fb059ca89f80ea095d2d)

commit 892883806f8d916f037ecc125ac7bc66090f6c81
Merge: 918f0e810 9e60b3ba4
Author: android-build-team Robot <android-build-team-robot@google.com>
Date:   Tue Aug 21 03:03:01 2018 +0000

    Snap for 4966094 from 9e60b3ba4922c7c2c85f0c111a7585002e7a8545 to pi-qpr1-release

    Change-Id: I0e44eb1c96c1465085b279abee251100d54cc0f1

commit 9e60b3ba4922c7c2c85f0c111a7585002e7a8545
Merge: 878fb46a5 635a1f1b1
Author: Naina Nalluri <nainanalluri@google.com>
Date:   Mon Aug 20 22:33:52 2018 +0000

    Merge "Store config file with gid1" into pi-dev

commit 918f0e8104114fc9dc318b2c9168a2e4770c2c1e
Merge: 357a7710b 878fb46a5
Author: android-build-team Robot <android-build-team-robot@google.com>
Date:   Fri Aug 17 03:00:59 2018 +0000

    Snap for 4959688 from 878fb46a571835137e32a163daaa68a8690ebd2f to pi-qpr1-release

    Change-Id: I6e798b20913034e4932e0eea3dd089364961eb68

commit 635a1f1b1bc05b0f272aebfcc755d36ddfca8760
Author: Naina Nalluri <nainanalluri@google.com>
Date:   Fri Aug 3 10:38:30 2018 -0700

    Store config file with gid1

    Store and look up for config file with iccid
    and gid1.

    Bug: 111755653
    Bug: 73538013
    Test: Carrier config loading on various carriers

    Change-Id: If349bbf2b7bad42738d8b08f3c7e8365062e55b3

commit 357a7710bafe4d478102864cd913bd2cf6a122cb
Merge: db6046365 f005ff9fd
Author: android-build-team Robot <android-build-team-robot@google.com>
Date:   Fri Jul 27 03:13:06 2018 +0000

    Snap for 4916303 from f005ff9fd1882cbbff50ed69c4509fcc75e57cb4 to pi-qpr1-release

    Change-Id: Ifb87e4138885f4f5d39ea8363529911920f46e80

commit 878fb46a571835137e32a163daaa68a8690ebd2f
Author: Hall Liu <hallliu@google.com>
Date:   Wed Jul 25 18:49:59 2018 -0700

    Add plumbing for the 911 redial

    Provide a mechanism to swap out the original connection when the lower
    layers trigger a redial.

    Bug: 111235453
    Test: manual (modify IMS stack locally to always return the 380 error
    when dialing in airplane mode)

    Change-Id: Ie1cab49f040dd2b34b775bd9e5d7f97fbf7aafa9

commit f005ff9fd1882cbbff50ed69c4509fcc75e57cb4
Author: Cassie <xih@google.com>
Date:   Mon Jul 16 13:07:59 2018 -0700

    Clean cherry-pick 'Handles the case when getForbiddenPlmns() returns null.'

    Bug: 111490801
    Test: Manual test
    Change-Id: I1dafdbb874468a8de15f7fad6de5a1478592c1ba
    Merged-In: I1dafdbb874468a8de15f7fad6de5a1478592c1ba

commit 38fb17b4a2517bd436a30b22cddde9974e6cdb9a
Merge: fb3d7c164 3732c2bc6
Author: Xiangyu/Malcolm Chen <refuhoo@google.com>
Date:   Fri Jul 20 21:36:04 2018 +0000

    Merge "Fix the bug TM#getNetworkCountryISO returns null in SUW." into pi-dev

commit fb3d7c164726d20a3847b7bf9ec1be610751823b
Author: Bill Yi <byi@google.com>
Date:   Thu Jul 19 21:39:57 2018 -0700

    Import translations. DO NOT MERGE

    Auto-generated-cl: translation import

    Bug: 64712476
    Change-Id: I2f2477778009ec18949b3157377bda8339e62588

commit 3732c2bc6d2e709c7727445059310184d5e1aeca
Author: Malcolm Chen <refuhoo@google.com>
Date:   Wed Jul 18 20:15:24 2018 -0700

    Fix the bug TM#getNetworkCountryISO returns null in SUW.

    TM#getNetworkCountryISO() will pass INVALID_PHONE_INDEX if no default
    SIM is ever set, and ends up having a null phone object. Then an
    empty string is returned instead of proper value from LocaleTracker#
    getCurrentCountry.

    Bug: 111387352
    Test: TelephonyManagerTestApp. With no SIM, call TM#getNetworkCountryISO
    and see the result.
    Change-Id: I2bf04620d58a9610b1b2ea53be193061350d0137

commit db6046365c3ff4929f638296bca853adefd98266
Merge: fe67097d9 d36c5e4a9
Author: Tyler Gunn <tgunn@google.com>
Date:   Mon Jul 16 19:19:24 2018 +0000

    Merge "Inform user to turn off APM for IMS_SIP_ALTERNATE_EMERGENCY_CALL cause." into pi-dev

commit d36c5e4a9b90446fb2ef6f7b7712176c4b732b32
Author: Tyler Gunn <tgunn@google.com>
Date:   Wed Jul 11 11:49:19 2018 -0700

    Inform user to turn off APM for IMS_SIP_ALTERNATE_EMERGENCY_CALL cause.

    When IMS_SIP_ALTERNATE_EMERGENCY_CALL disconnect cause is received,
    inform the user to turn off airplane mode in order to make a call.

    Test: make -j 64; no manual testing possible.
    Bug: 69024173
    Change-Id: I76afa4796f1b7ad2b9595bc8680559fa9e83d462

commit fe67097d9159a0a5be3c3d2f09e4709c3f94e496
Merge: 790e02ac3 a18d19b5d
Author: Bill Yi <byi@google.com>
Date:   Fri Jul 6 01:30:19 2018 -0700

    [automerger skipped] Import translations. DO NOT MERGE skipped: ed887297b9 skipped: 594624f336 am: a9ef860b83  -s ours am: 0e9f6c0928  -s ours
    am: a18d19b5d6  -s ours

    Change-Id: I69242979534bd2ebf69a58e6987e7c3eef9bef52

commit a18d19b5d66b5dd5eb9e6a12890bdbca83d56fdc
Merge: bc96cc193 0e9f6c092
Author: Bill Yi <byi@google.com>
Date:   Fri Jul 6 00:24:51 2018 -0700

    [automerger skipped] Import translations. DO NOT MERGE skipped: ed887297b9 skipped: 594624f336 am: a9ef860b83  -s ours
    am: 0e9f6c0928  -s ours

    Change-Id: Ib5b7491931d6178bd5dfe7eb3844754d2a84bb7d

commit 0e9f6c0928e3868035a16779102d79b4992759b9
Merge: 77ce4e9cb a9ef860b8
Author: Bill Yi <byi@google.com>
Date:   Thu Jul 5 21:05:04 2018 -0700

    [automerger skipped] Import translations. DO NOT MERGE skipped: ed887297b9 skipped: 594624f336
    am: a9ef860b83  -s ours

    Change-Id: I8d1583d27fe4473fed6df79b161af1c25a2c8566

commit a9ef860b8310410aaee399fcc1c032b9a2267250
Merge: d3579f0c7 594624f33
Author: Android Build Merger (Role) <noreply-android-build-merger@google.com>
Date:   Thu Jul 5 08:26:55 2018 +0000

    [automerger skipped] Import translations. DO NOT MERGE skipped: ed887297b9 skipped: 594624f336

    Change-Id: I3069820676a7f548bca01669b91cef92dbcae319

commit 594624f3369538c1a002ec687684cf0b628f688b
Merge: 89b1eb637 ed887297b
Author: Android Build Merger (Role) <noreply-android-build-merger@google.com>
Date:   Thu Jul 5 08:26:53 2018 +0000

    [automerger skipped] Import translations. DO NOT MERGE skipped: ed887297b9

    Change-Id: I6bb5f60ed63c96fec695f4223a6212340bf0775b

commit ed887297b9355159632e124194b39d09347e9cd0
Author: Bill Yi <byi@google.com>
Date:   Thu Jul 5 01:26:41 2018 -0700

    Import translations. DO NOT MERGE

    Change-Id: I6393e57f2cc5720c795aa68a1faea55edb06c0a6
    Auto-generated-cl: translation import

commit 790e02ac339b0adb9ab1afb69cbcf79fba9dfdb8
Author: Bill Yi <byi@google.com>
Date:   Thu Jul 5 01:16:40 2018 -0700

    Import translations. DO NOT MERGE

    Auto-generated-cl: translation import

    Bug: 64712476
    Change-Id: I6b748b3cbea5915a7cb5598b80ba143f3bb338d8

commit 35d17261e5f69d6d63fd78e962787c9cc6697b8c
Merge: f6fd992fe 322cccb9d
Author: TreeHugger Robot <treehugger-gerrit@google.com>
Date:   Tue Jul 3 00:01:52 2018 +0000

    Merge "Show or hide call forwarding,additional settings according to carrier config." into pi-dev

commit f6fd992fe7adb530013f5d28d0bd32aa401ad4d4
Author: SongFerngWang <songferngwang@google.com>
Date:   Thu Jun 28 17:21:43 2018 +0800

    Show or hide call forwarding,additional settings according to carrier config.

    Show or hide call forwarding,additional settings,clir and call waiting according to
    KEY_CALL_FORWARDING_VISIBILITY_BOOL, KEY_ADDITIONAL_SETTINGS_CALLER_ID_VISIBILITY_BOOL
    and KEY_ADDITIONAL_SETTINGS_CALL_WAITING_VISIBILITY_BOOL.

    Bug: 110121209
    Test: manual - test case as below :
    1.flag clir on and flag cw on ,do CLIR show/hide ,
        and make call to other device.(PASS）
    2.flag clir on and flag cw on ,do CW on/off,and CW function work.（PASS）
    3.flag clir off and flag cw on ,no CLIR in additional settings page（PASS）
    4.flag clir off and flag cw on ,do CW on/off,and CW function work.（PASS）
    5.flag clir on and flag cw off ,do CLIR show/hide ,
        and make call to other device.（PASS）
    6.flag clir on and flag cw off ,no CW in additional settings page（PASS）
    7.flag clir off and flag cw off,
        no Additional Settings in call feature settings（PASS）
    8.flag cf off, no CF in call feature settings (PASS)

    Change-Id: I4f57438e9776f182812614d66a40ed25d3d73e9d

commit 322cccb9d524a8f88704a80505473c25cffeb014
Author: SongFerngWang <songferngwang@google.com>
Date:   Fri Jun 29 15:47:58 2018 +0800

    Show or hide call forwarding,additional settings according to carrier config.

    Show or hide call forwarding,additional settings,clir and call waiting according to
    KEY_CALL_FORWARDING_VISIBILITY_BOOL, KEY_ADDITIONAL_SETTINGS_CALLER_ID_VISIBILITY_BOOL
    and KEY_ADDITIONAL_SETTINGS_CALL_WAITING_VISIBILITY_BOOL.

    Bug: 110121209
    Test: manual - test case as below :
    1.flag clir on and flag cw on ,do CLIR show/hide ,
        and make call to other device.(PASS）
    2.flag clir on and flag cw on ,do CW on/off,and CW function work.（PASS）
    3.flag clir off and flag cw on ,no CLIR in additional settings page（PASS）
    4.flag clir off and flag cw on ,do CW on/off,and CW function work.（PASS）
    5.flag clir on and flag cw off ,do CLIR show/hide ,
        and make call to other device.（PASS）
    6.flag clir on and flag cw off ,no CW in additional settings page（PASS）
    7.flag clir off and flag cw off,
        no Additional Settings in call feature settings（PASS）
    8.flag cf off, no CF in call feature settings (PASS)

    Change-Id: If74f56b9f99d1273b2cd7cd9e159fc47eca878c7

commit 1231b7016e3d0107db6abef925967e5338db7b73
Merge: 90a4fbf86 f5d29555b
Author: TreeHugger Robot <treehugger-gerrit@google.com>
Date:   Mon Jun 25 19:47:26 2018 +0000

    Merge "Fix in Modem Activity Info updates" into pi-dev

commit f5d29555b53457e8172552dbb0228bf178bbd4a8
Author: Siddharth Ray <siddharthr@google.com>
Date:   Sun Jun 17 15:02:38 2018 -0700

    Fix in Modem Activity Info updates

    A check is added to capture the following error conditions:
    - Idle time is larger than the total activity duration.
    - Sleep time is larger than the total activity duration.
    - Rx time is larger than the total activity duration.
    - Total Tx time is larger than the total activity duration.

    BUG:110319204
    Change-Id: I47b18313c0d0e336110859bd79da2347862bbfde
    Merged-In: I47b18313c0d0e336110859bd79da2347862bbfde

commit 90a4fbf860e006fdf1658a49bc3e83c2a2af4099
Author: Bill Yi <byi@google.com>
Date:   Thu Jun 21 15:02:33 2018 -0700

    Import translations. DO NOT MERGE

    Auto-generated-cl: translation import

    Bug: 64712476
    Change-Id: Id0c0353ab2baac0b2a9fd41c7a8390273c7c3374

commit ddaeb3bf9e27af1d841d92908eda1e5c5e4f21e2
Author: Brad Ebinger <breadley@google.com>
Date:   Wed Jun 13 17:27:12 2018 -0700

    Revert isResolvingImsBinding implementation

    Bug: 109762920
    Test: Telephony GTS/Unit Testing
    Change-Id: I9040634407f574112d2458c5c8e454b9316705e5
    Merged-In: I9040634407f574112d2458c5c8e454b9316705e5

commit 050230da76f3ebcfd94b3dc93109638574b09757
Merge: 0617db0a2 067853f13
Author: TreeHugger Robot <treehugger-gerrit@google.com>
Date:   Mon Jun 18 22:27:55 2018 +0000

    Merge "Move receiver registration to onPause/onResume for MobileNetworkSettings." into pi-dev

commit 0617db0a21146d91760f978c97ed318d4b43f2dd
Author: Hall Liu <hallliu@google.com>
Date:   Tue Jun 12 11:29:32 2018 -0700

    Mark RTT calls as non-conferenceable

    In refreshConferenceSupported, check if the call is RTT and mark it
    non-conferenceable if so. Also call refreshConferenceSupported when the
    RTT status changes.

    Bug: 109763648
    Test: manual
    Change-Id: I374b737d4f5dabec3091fb2da2eaf8355d54d221

commit 067853f132286e03eda8df0e19b197a6f9f8a4c8
Author: Tyler Gunn <tgunn@google.com>
Date:   Tue Jun 12 09:38:54 2018 -0700

    Move receiver registration to onPause/onResume for MobileNetworkSettings.

    Moving registration and unregistration of receivers on the
    MobileNetworkSettings page to the onPause/onResume methods.
    Previously this was done in onCreate and onDestroy.
    Also, adding null checks in the receivers to ensure that they won't try
    to update the UI when the activity is null.

    Test: Manual
    Bug: 109899326
    Change-Id: Id4d6c37d155824b26ee4c3dd33bb6bd8ac7b619b
    Merged-In: Id4d6c37d155824b26ee4c3dd33bb6bd8ac7b619b
    (cherry picked from commit c3dbff7ccd08a6acb34de9df95424a2d30deaaad)

commit 89b1eb637307e985f6a1a644268664ec60f444af
Merge: f45b5b3cb 03900d8e8
Author: Tyler Gunn <tgunn@google.com>
Date:   Tue Jun 5 16:40:04 2018 +0000

    Merge changes from topic "am-538bf8fd-0fa8-4c10-aeea-2cd19adfb09f" into nyc-mr2-dev

    * changes:
      [automerger] DO NOT MERGE Revise security requirements for TelephonyManager#endCall. am: 5ae9c95f56 am: 91971125eb am: e89a641900 am: ee4c0fedba
      [automerger] DO NOT MERGE Revise security requirements for TelephonyManager#endCall. am: 5ae9c95f56 am: 91971125eb am: e89a641900
      [automerger] DO NOT MERGE Revise security requirements for TelephonyManager#endCall. am: 5ae9c95f56 am: 91971125eb
      [automerger] DO NOT MERGE Revise security requirements for TelephonyManager#endCall. am: 5ae9c95f56
      DO NOT MERGE Revise security requirements for TelephonyManager#endCall.

commit 16c1a7e30d8eb87d54eb2ccc022cfc541e8f5542
Merge: 6893327c6 ee4c0fedb
Author: Tyler Gunn <tgunn@google.com>
Date:   Tue Jun 5 16:40:04 2018 +0000

    Merge changes from topic "am-538bf8fd-0fa8-4c10-aeea-2cd19adfb09f" into cw-f-dev

    * changes:
      [automerger] DO NOT MERGE Revise security requirements for TelephonyManager#endCall. am: 5ae9c95f56 am: 91971125eb am: e89a641900
      [automerger] DO NOT MERGE Revise security requirements for TelephonyManager#endCall. am: 5ae9c95f56 am: 91971125eb
      [automerger] DO NOT MERGE Revise security requirements for TelephonyManager#endCall. am: 5ae9c95f56
      DO NOT MERGE Revise security requirements for TelephonyManager#endCall.

Change-Id: Ia2d7b9b860d454d145c10b7674b0af054884b112
---
 res/values-da/strings.xml                     |  6 +-
 res/values-de/strings.xml                     |  4 +-
 res/values-gl/strings.xml                     |  4 +-
 res/values-hi/strings.xml                     |  2 +-
 res/values-ne/strings.xml                     |  2 +-
 res/values-sw/strings.xml                     |  4 +-
 sip/res/values-iw/strings.xml                 |  2 +-
 .../android/phone/CarrierConfigLoader.java    | 57 +++++++++---
 .../phone/GsmUmtsAdditionalCallOptions.java   | 91 ++++++++++++++-----
 src/com/android/phone/GsmUmtsCallOptions.java | 48 +++++++---
 .../android/phone/MobileNetworkSettings.java  | 31 ++++---
 .../phone/NetworkSelectListPreference.java    |  3 +-
 .../android/phone/NetworkSelectSetting.java   |  3 +-
 .../android/phone/PhoneInterfaceManager.java  | 34 +++++--
 .../telephony/TelecomAccountRegistry.java     | 86 +++++++++++++++---
 .../telephony/TelephonyConnection.java        | 17 +++-
 16 files changed, 295 insertions(+), 99 deletions(-)

diff --git a/res/values-da/strings.xml b/res/values-da/strings.xml
index f0dcb67c8..c71dc81c2 100644
--- a/res/values-da/strings.xml
+++ b/res/values-da/strings.xml
@@ -417,7 +417,7 @@
     <string name="menu_add" msgid="1882023737425114762">"Tilføj kontakt"</string>
     <string name="menu_edit" msgid="7143003705504672374">"Rediger kontakt"</string>
     <string name="menu_delete" msgid="3977150783449642851">"Slet kontakt"</string>
-    <string name="menu_dial" msgid="3223106222819685808">"Ring til kontaktperson"</string>
+    <string name="menu_dial" msgid="3223106222819685808">"Ring til kontakt"</string>
     <string name="get_pin2" msgid="8204677063922225311">"Angiv PIN2-kode"</string>
     <string name="name" msgid="7329028332786872378">"Navn"</string>
     <string name="number" msgid="7905950798349903858">"Nummer"</string>
@@ -596,9 +596,9 @@
     <string name="voicemail_number_not_set" msgid="6724904736891087856">"&lt;Ikke angivet&gt;"</string>
     <string name="other_settings" msgid="3672912580359716394">"Andre indstillinger for opkald"</string>
     <string name="calling_via_template" msgid="4839419581866928142">"Opkald via <xliff:g id="PROVIDER_NAME">%s</xliff:g>"</string>
-    <string name="contactPhoto" msgid="4713193418046639466">"billede af kontaktperson"</string>
+    <string name="contactPhoto" msgid="4713193418046639466">"billede af kontakt"</string>
     <string name="goPrivate" msgid="865837794424530980">"gør privat"</string>
-    <string name="selectContact" msgid="781975788478987237">"vælg kontaktperson"</string>
+    <string name="selectContact" msgid="781975788478987237">"vælg kontakt"</string>
     <string name="not_voice_capable" msgid="2739898841461577811">"Taleopkald understøttes ikke"</string>
     <string name="description_dial_button" msgid="7459705245418435351">"ring op"</string>
     <string name="voicemail_visual_voicemail_switch_title" msgid="5012622186976275457">"Visuel telefonsvarer"</string>
diff --git a/res/values-de/strings.xml b/res/values-de/strings.xml
index 17012442d..0b364318b 100644
--- a/res/values-de/strings.xml
+++ b/res/values-de/strings.xml
@@ -376,9 +376,9 @@
     <item msgid="176474317493999285">"Nur Heimatnetz"</item>
     <item msgid="1205664026446156265">"Automatisch"</item>
   </string-array>
-    <string name="cdma_subscription_title" msgid="1162564010076763284">"CDMA-Abonnement"</string>
+    <string name="cdma_subscription_title" msgid="1162564010076763284">"CDMA-Abo"</string>
     <string name="cdma_subscription_summary" msgid="2530890766115781140">"Zwischen RUIM/SIM und NV wechseln"</string>
-    <string name="cdma_subscription_dialogtitle" msgid="2699527950523333110">"Abonnement"</string>
+    <string name="cdma_subscription_dialogtitle" msgid="2699527950523333110">"Abo"</string>
   <string-array name="cdma_subscription_choices">
     <item msgid="2258014151300708431">"RUIM/SIM"</item>
     <item msgid="5127722015571873880">"NV"</item>
diff --git a/res/values-gl/strings.xml b/res/values-gl/strings.xml
index 75b42f5a4..236362fbd 100644
--- a/res/values-gl/strings.xml
+++ b/res/values-gl/strings.xml
@@ -327,8 +327,8 @@
   </string-array>
     <string name="list_language_dtitle" msgid="5442908726538951934">"Idiomas"</string>
     <string name="enable_disable_local_weather" msgid="986967454867219114">"Tempo local"</string>
-    <string name="local_weather_enable" msgid="6199315114382448922">"Información meteorolóxica local activada"</string>
-    <string name="local_weather_disable" msgid="2510158089142626480">"Información meteorolóxica local desactivada"</string>
+    <string name="local_weather_enable" msgid="6199315114382448922">"O tempo local activado"</string>
+    <string name="local_weather_disable" msgid="2510158089142626480">"O tempo local desactivado"</string>
     <string name="enable_disable_atr" msgid="8339572391278872343">"Informes de tráfico de zona"</string>
     <string name="atr_enable" msgid="5541757457789181799">"Informes de tráfico de zona activados"</string>
     <string name="atr_disable" msgid="7085558154727596455">"Informes de tráfico de zona desactivados"</string>
diff --git a/res/values-hi/strings.xml b/res/values-hi/strings.xml
index eeeef063f..8f41e8fdb 100644
--- a/res/values-hi/strings.xml
+++ b/res/values-hi/strings.xml
@@ -532,7 +532,7 @@
     <string name="hac_mode_summary" msgid="6833851160514929341">"श्रवण सहायक साधन संगतता चालू करें"</string>
     <string name="rtt_mode_title" msgid="6954652435543570121">"(आरटीटी) कॉल से रीयल-टाइम में मैसेज भेजें"</string>
     <string name="rtt_mode_summary" msgid="5146344277246544580">"वॉइस कॉल में मैसेज भेजने की अनुमति दें"</string>
-    <string name="rtt_mode_more_information" msgid="6080420780730383030">"आरटीटी उन कॉल करने वालों की मदद करता है, जो बधिर हैं, ऊंचा सुनते हैं, जिन्हें बोलने में दिक्कत है या जिनके लिए सिर्फ़ आवाज़ काफ़ी नहीं होती.&lt;br&gt; &lt;a href=<xliff:g id="URL">http://support.google.com/mobile?p=telephony_rtt</xliff:g>&gt;अधिक जानें&lt;/a&gt;\n       &lt;br&gt;&lt;br&gt; - आरटीटी कॉल, मैसेज लेख के रूप में सेव किए जाते हैं\n       &lt;br&gt; - वीडियो कॉल के लिए आरटीटी उपलब्ध नहीं है"</string>
+    <string name="rtt_mode_more_information" msgid="6080420780730383030">"आरटीटी उन कॉल करने वालों की मदद करता है, जो सुन नहीं सकते हैं, ऊंचा सुनते हैं, जिन्हें बोलने में दिक्कत है या जिनके लिए सिर्फ़ आवाज़ काफ़ी नहीं होती.&lt;br&gt; &lt;a href=<xliff:g id="URL">http://support.google.com/mobile?p=telephony_rtt</xliff:g>&gt;ज़्यादा  जानें&lt;/a&gt;\n       &lt;br&gt;&lt;br&gt; - आरटीटी कॉल, मैसेज लेख के रूप में सेव किए जाते हैं\n       &lt;br&gt; - वीडियो कॉल के लिए आरटीटी उपलब्ध नहीं है"</string>
   <string-array name="tty_mode_entries">
     <item msgid="512950011423868021">"TTY बंद"</item>
     <item msgid="3971695875449640648">"TTY पूर्ण"</item>
diff --git a/res/values-ne/strings.xml b/res/values-ne/strings.xml
index 6d473ce7b..6d6abe00f 100644
--- a/res/values-ne/strings.xml
+++ b/res/values-ne/strings.xml
@@ -136,7 +136,7 @@
     <string name="close_dialog" msgid="2365884406356986917">"ठिक छ"</string>
     <string name="enable" msgid="7248657275000173526">"सकृय पार्नुहोस्"</string>
     <string name="disable" msgid="4678348128118573672">"निस्कृय पार्नुहोस्"</string>
-    <string name="change_num" msgid="239476305819844391">"अपडेट गर्नुहोस्"</string>
+    <string name="change_num" msgid="239476305819844391">"अद्यावधिक गर्नुहोस्"</string>
   <string-array name="clir_display_values">
     <item msgid="5560134294467334594">"नेटवर्क पूर्वनिर्धारित"</item>
     <item msgid="7876195870037833661">"सङ्ख्या लुकाउनुहोस्"</item>
diff --git a/res/values-sw/strings.xml b/res/values-sw/strings.xml
index 8e07fe3c3..606231335 100644
--- a/res/values-sw/strings.xml
+++ b/res/values-sw/strings.xml
@@ -90,7 +90,7 @@
     <string name="sum_loading_settings" msgid="1826692909391168620">"Mipangilio inapakia..."</string>
     <string name="sum_hide_caller_id" msgid="1071407020290873782">"Nambari imefichwa kwa simu unayopiga"</string>
     <string name="sum_show_caller_id" msgid="6768534125447290401">"Namba inaonekana kwa simu zinazopigwa"</string>
-    <string name="sum_default_caller_id" msgid="1954518825510901365">"Tumia mipangilio ya kiendesha chaguo msingi kuonyesha namba kwa simu zinazopigwa"</string>
+    <string name="sum_default_caller_id" msgid="1954518825510901365">"Tumia mipangilio ya kiendesha chaguomsingi kuonyesha namba kwa simu zinazopigwa"</string>
     <string name="labelCW" msgid="6120513814915920200">"Simu inayosubiri kupokewa"</string>
     <string name="sum_cw_enabled" msgid="8083061901633671397">"Wakati ninapokea simu, niarifu kuhusu simu zingine zinazoingia"</string>
     <string name="sum_cw_disabled" msgid="3648693907300104575">"Wakati ninapokea simu, niarifu kuhusu simu zingine zinazoingia"</string>
@@ -138,7 +138,7 @@
     <string name="disable" msgid="4678348128118573672">"Zima"</string>
     <string name="change_num" msgid="239476305819844391">"Sasisha"</string>
   <string-array name="clir_display_values">
-    <item msgid="5560134294467334594">"Chaguo msingi la mtandao"</item>
+    <item msgid="5560134294467334594">"Chaguomsingi la mtandao"</item>
     <item msgid="7876195870037833661">"Ficha nambari"</item>
     <item msgid="1108394741608734023">"Onyesha nambari"</item>
   </string-array>
diff --git a/sip/res/values-iw/strings.xml b/sip/res/values-iw/strings.xml
index edd8d649d..2204e0ac0 100644
--- a/sip/res/values-iw/strings.xml
+++ b/sip/res/values-iw/strings.xml
@@ -26,7 +26,7 @@
     <string name="sip_call_options_entry_1" msgid="7217659161237099900">"לכל השיחות כאשר רשת הנתונים זמינה"</string>
     <string name="sip_call_options_entry_2" msgid="1815335903940609729">"‏עבור שיחות SIP בלבד"</string>
     <string name="sip_call_options_wifi_only_entry_1" msgid="1358513095551847314">"לכל השיחות"</string>
-    <string name="add_sip_account" msgid="800843669753980091">"הוסף חשבון"</string>
+    <string name="add_sip_account" msgid="800843669753980091">"חשבון חדש"</string>
     <string name="remove_sip_account" msgid="1367664438506503690">"הסר חשבון"</string>
     <string name="sip_account_list" msgid="5610858485304821480">"‏חשבונות SIP"</string>
     <string name="saving_account" msgid="5336529880235177448">"שומר את החשבון..."</string>
diff --git a/src/com/android/phone/CarrierConfigLoader.java b/src/com/android/phone/CarrierConfigLoader.java
index 0424a031f..f216b7940 100644
--- a/src/com/android/phone/CarrierConfigLoader.java
+++ b/src/com/android/phone/CarrierConfigLoader.java
@@ -72,6 +72,7 @@ import java.io.FileOutputStream;
 import java.io.FilenameFilter;
 import java.io.IOException;
 import java.io.PrintWriter;
+import java.lang.Math;
 import java.util.ArrayList;
 import java.util.Arrays;
 import java.util.Collections;
@@ -137,6 +138,9 @@ public class CarrierConfigLoader extends ICarrierConfigLoader.Stub {
 
     private static final int BIND_TIMEOUT_MILLIS = 30000;
 
+    // Length limit of gid1 for naming config file.
+    private static final int GID1_LENGTH_LIMIT = 20;
+
     // Tags used for saving and restoring XML documents.
     private static final String TAG_DOCUMENT = "carrier_config";
     private static final String TAG_VERSION = "package_version";
@@ -218,8 +222,9 @@ public class CarrierConfigLoader extends ICarrierConfigLoader.Stub {
                 case EVENT_DO_FETCH_DEFAULT:
                 {
                     final String iccid = getIccIdForPhoneId(phoneId);
+                    final String gid1 = getGid1ForPhoneId(phoneId);
                     final PersistableBundle config =
-                            restoreConfigFromXml(mPlatformCarrierConfigPackage, iccid);
+                            restoreConfigFromXml(mPlatformCarrierConfigPackage, iccid, gid1);
                     if (config != null) {
                         log(
                                 "Loaded config from XML. package="
@@ -260,6 +265,7 @@ public class CarrierConfigLoader extends ICarrierConfigLoader.Stub {
                     }
                     final CarrierIdentifier carrierId = getCarrierIdForPhoneId(phoneId);
                     final String iccid = getIccIdForPhoneId(phoneId);
+                    final String gid1 = getGid1ForPhoneId(phoneId);
                     // ResultReceiver callback will execute in this Handler's thread.
                     final ResultReceiver resultReceiver =
                             new ResultReceiver(this) {
@@ -280,8 +286,8 @@ public class CarrierConfigLoader extends ICarrierConfigLoader.Stub {
                                     }
                                     PersistableBundle config =
                                             resultData.getParcelable(KEY_CONFIG_BUNDLE);
-                                    saveConfigToXml(
-                                            mPlatformCarrierConfigPackage, iccid, config);
+                                    saveConfigToXml(mPlatformCarrierConfigPackage,
+                                            iccid, gid1, config);
                                     mConfigFromDefaultApp[phoneId] = config;
                                     sendMessage(
                                             obtainMessage(
@@ -337,8 +343,9 @@ public class CarrierConfigLoader extends ICarrierConfigLoader.Stub {
                 {
                     final String carrierPackageName = getCarrierPackageForPhoneId(phoneId);
                     final String iccid = getIccIdForPhoneId(phoneId);
+                    final String gid1 = getGid1ForPhoneId(phoneId);
                     final PersistableBundle config =
-                            restoreConfigFromXml(carrierPackageName, iccid);
+                            restoreConfigFromXml(carrierPackageName, iccid, gid1);
                     if (config != null) {
                         log(
                                 "Loaded config from XML. package="
@@ -378,6 +385,7 @@ public class CarrierConfigLoader extends ICarrierConfigLoader.Stub {
                     }
                     final CarrierIdentifier carrierId = getCarrierIdForPhoneId(phoneId);
                     final String iccid = getIccIdForPhoneId(phoneId);
+                    final String gid1 = getGid1ForPhoneId(phoneId);
                     // ResultReceiver callback will execute in this Handler's thread.
                     final ResultReceiver resultReceiver =
                             new ResultReceiver(this) {
@@ -398,8 +406,8 @@ public class CarrierConfigLoader extends ICarrierConfigLoader.Stub {
                                     }
                                     PersistableBundle config =
                                             resultData.getParcelable(KEY_CONFIG_BUNDLE);
-                                    saveConfigToXml(
-                                            getCarrierPackageForPhoneId(phoneId), iccid, config);
+                                    saveConfigToXml(getCarrierPackageForPhoneId(phoneId),
+                                            iccid, gid1, config);
                                     mConfigFromCarrierApp[phoneId] = config;
                                     sendMessage(
                                             obtainMessage(
@@ -601,6 +609,21 @@ public class CarrierConfigLoader extends ICarrierConfigLoader.Stub {
         return phone.getIccSerialNumber();
     }
 
+    private String getGid1ForPhoneId(int phoneId) {
+        if (!SubscriptionManager.isValidPhoneId(phoneId)) {
+            return null;
+        }
+        Phone phone = PhoneFactory.getPhone(phoneId);
+        if (phone == null) {
+            return null;
+        }
+        String gid1 = phone.getGroupIdLevel1();
+        if (gid1 == null) {
+            return null;
+        }
+        return gid1.substring(0, Math.min(gid1.length(), GID1_LENGTH_LIMIT));
+    }
+
     /**
      * Writes a bundle to an XML file.
      *
@@ -612,9 +635,12 @@ public class CarrierConfigLoader extends ICarrierConfigLoader.Stub {
      *
      * @param packageName the name of the package from which we fetched this bundle.
      * @param iccid the ICCID of the subscription for which this bundle was fetched.
+     * @param extras First 20 characters of gid1 of the subscription for which the bundle
+     *               was fetched.
      * @param config the bundle to be written. Null will be treated as an empty bundle.
      */
-    private void saveConfigToXml(String packageName, String iccid, PersistableBundle config) {
+    private void saveConfigToXml(String packageName, String iccid, String extras,
+            PersistableBundle config) {
         if (packageName == null || iccid == null) {
             loge("Cannot save config with null packageName or iccid.");
             return;
@@ -637,7 +663,8 @@ public class CarrierConfigLoader extends ICarrierConfigLoader.Stub {
         FileOutputStream outFile = null;
         try {
             outFile = new FileOutputStream(
-                    new File(mContext.getFilesDir(), getFilenameForConfig(packageName, iccid)));
+                    new File(mContext.getFilesDir(), getFilenameForConfig(packageName,
+                            iccid, extras)));
             FastXmlSerializer out = new FastXmlSerializer();
             out.setOutput(outFile, "utf-8");
             out.startDocument("utf-8", true);
@@ -672,10 +699,13 @@ public class CarrierConfigLoader extends ICarrierConfigLoader.Stub {
      *
      * @param packageName the name of the package from which we fetched this bundle.
      * @param iccid the ICCID of the subscription for which this bundle was fetched.
+     * @param extras First 20 characters of gid1 of the subscription for which the bundle
+     *               was fetched.
      * @return the bundle from the XML file. Returns null if there is no saved config, the saved
      *         version does not match, or reading config fails.
      */
-    private PersistableBundle restoreConfigFromXml(String packageName, String iccid) {
+    private PersistableBundle restoreConfigFromXml(String packageName, String iccid,
+            String extras) {
         final String version = getPackageVersion(packageName);
         if (version == null) {
             loge("Failed to get package version for: " + packageName);
@@ -690,7 +720,8 @@ public class CarrierConfigLoader extends ICarrierConfigLoader.Stub {
         FileInputStream inFile = null;
         try {
             inFile = new FileInputStream(
-                    new File(mContext.getFilesDir(), getFilenameForConfig(packageName, iccid)));
+                    new File(mContext.getFilesDir(), getFilenameForConfig(packageName, iccid,
+                            extras)));
             XmlPullParser parser = XmlPullParserFactory.newInstance().newPullParser();
             parser.setInput(inFile, "utf-8");
 
@@ -753,7 +784,11 @@ public class CarrierConfigLoader extends ICarrierConfigLoader.Stub {
     }
 
     /** Builds a canonical file name for a config file. */
-    private String getFilenameForConfig(@NonNull String packageName, @NonNull String iccid) {
+    private String getFilenameForConfig(@NonNull String packageName, @NonNull String iccid,
+            String extras) {
+        if (extras != null) {
+            return "carrierconfig-" + packageName + "-" + iccid + "-" + extras + ".xml";
+        }
         return "carrierconfig-" + packageName + "-" + iccid + ".xml";
     }
 
diff --git a/src/com/android/phone/GsmUmtsAdditionalCallOptions.java b/src/com/android/phone/GsmUmtsAdditionalCallOptions.java
index 082eaa9e2..b79cdd8f7 100644
--- a/src/com/android/phone/GsmUmtsAdditionalCallOptions.java
+++ b/src/com/android/phone/GsmUmtsAdditionalCallOptions.java
@@ -2,8 +2,10 @@ package com.android.phone;
 
 import android.app.ActionBar;
 import android.os.Bundle;
+import android.os.PersistableBundle;
 import android.preference.Preference;
 import android.preference.PreferenceScreen;
+import android.telephony.CarrierConfigManager;
 import android.util.Log;
 import android.view.MenuItem;
 
@@ -26,6 +28,9 @@ public class GsmUmtsAdditionalCallOptions extends TimeConsumingPreferenceActivit
     private Phone mPhone;
     private SubscriptionInfoHelper mSubscriptionInfoHelper;
 
+    private boolean mShowCLIRButton;
+    private boolean mShowCWButton;
+
     @Override
     protected void onCreate(Bundle icicle) {
         super.onCreate(icicle);
@@ -41,24 +46,60 @@ public class GsmUmtsAdditionalCallOptions extends TimeConsumingPreferenceActivit
         mCLIRButton = (CLIRListPreference) prefSet.findPreference(BUTTON_CLIR_KEY);
         mCWButton = (CallWaitingSwitchPreference) prefSet.findPreference(BUTTON_CW_KEY);
 
-        mPreferences.add(mCLIRButton);
-        mPreferences.add(mCWButton);
-
-        if (icicle == null) {
-            if (DBG) Log.d(LOG_TAG, "start to init ");
-            mCLIRButton.init(this, false, mPhone);
+        PersistableBundle b = null;
+        if (mSubscriptionInfoHelper.hasSubId()) {
+            b = PhoneGlobals.getInstance().getCarrierConfigForSubId(
+                    mSubscriptionInfoHelper.getSubId());
         } else {
-            if (DBG) Log.d(LOG_TAG, "restore stored states");
-            mInitIndex = mPreferences.size();
-            mCLIRButton.init(this, true, mPhone);
-            mCWButton.init(this, true, mPhone);
-            int[] clirArray = icicle.getIntArray(mCLIRButton.getKey());
-            if (clirArray != null) {
-                if (DBG) Log.d(LOG_TAG, "onCreate:  clirArray[0]="
-                        + clirArray[0] + ", clirArray[1]=" + clirArray[1]);
-                mCLIRButton.handleGetCLIRResult(clirArray);
+            b = PhoneGlobals.getInstance().getCarrierConfig();
+        }
+
+        if (b != null) {
+            mShowCLIRButton = b.getBoolean(
+                    CarrierConfigManager.KEY_ADDITIONAL_SETTINGS_CALLER_ID_VISIBILITY_BOOL);
+            mShowCWButton = b.getBoolean(
+                    CarrierConfigManager.KEY_ADDITIONAL_SETTINGS_CALL_WAITING_VISIBILITY_BOOL);
+        }
+
+        if (mCLIRButton != null) {
+            if (mShowCLIRButton) {
+                mPreferences.add(mCLIRButton);
             } else {
-                mCLIRButton.init(this, false, mPhone);
+                prefSet.removePreference(mCLIRButton);
+            }
+        }
+
+        if (mCWButton != null) {
+            if (mShowCWButton) {
+                mPreferences.add(mCWButton);
+            } else {
+                prefSet.removePreference(mCWButton);
+            }
+        }
+
+        if (mPreferences.size() != 0) {
+            if (icicle == null) {
+                if (DBG) Log.d(LOG_TAG, "start to init ");
+                doPreferenceInit(mInitIndex);
+            } else {
+                if (DBG) Log.d(LOG_TAG, "restore stored states");
+                mInitIndex = mPreferences.size();
+                if (mShowCWButton) {
+                    mCWButton.init(this, true, mPhone);
+                }
+                if (mShowCLIRButton) {
+                    mCLIRButton.init(this, true, mPhone);
+                    int[] clirArray = icicle.getIntArray(mCLIRButton.getKey());
+                    if (clirArray != null) {
+                        if (DBG) {
+                            Log.d(LOG_TAG, "onCreate:  clirArray[0]="
+                                    + clirArray[0] + ", clirArray[1]=" + clirArray[1]);
+                        }
+                        mCLIRButton.handleGetCLIRResult(clirArray);
+                    } else {
+                        mCLIRButton.init(this, false, mPhone);
+                    }
+                }
             }
         }
 
@@ -73,7 +114,7 @@ public class GsmUmtsAdditionalCallOptions extends TimeConsumingPreferenceActivit
     protected void onSaveInstanceState(Bundle outState) {
         super.onSaveInstanceState(outState);
 
-        if (mCLIRButton.clirArray != null) {
+        if (mShowCLIRButton && mCLIRButton.clirArray != null) {
             outState.putIntArray(mCLIRButton.getKey(), mCLIRButton.clirArray);
         }
     }
@@ -82,10 +123,7 @@ public class GsmUmtsAdditionalCallOptions extends TimeConsumingPreferenceActivit
     public void onFinished(Preference preference, boolean reading) {
         if (mInitIndex < mPreferences.size()-1 && !isFinishing()) {
             mInitIndex++;
-            Preference pref = mPreferences.get(mInitIndex);
-            if (pref instanceof CallWaitingSwitchPreference) {
-                ((CallWaitingSwitchPreference) pref).init(this, false, mPhone);
-            }
+            doPreferenceInit(mInitIndex);
         }
         super.onFinished(preference, reading);
     }
@@ -99,4 +137,15 @@ public class GsmUmtsAdditionalCallOptions extends TimeConsumingPreferenceActivit
         }
         return super.onOptionsItemSelected(item);
     }
+
+    private void doPreferenceInit(int index) {
+        if (mPreferences.size() != 0) {
+            Preference pref = mPreferences.get(index);
+            if (pref instanceof CallWaitingSwitchPreference) {
+                ((CallWaitingSwitchPreference) pref).init(this, false, mPhone);
+            } else if (pref instanceof CLIRListPreference) {
+                ((CLIRListPreference) pref).init(this, false, mPhone);
+            }
+        }
+    }
 }
diff --git a/src/com/android/phone/GsmUmtsCallOptions.java b/src/com/android/phone/GsmUmtsCallOptions.java
index 3b27d2813..ab44b54bb 100644
--- a/src/com/android/phone/GsmUmtsCallOptions.java
+++ b/src/com/android/phone/GsmUmtsCallOptions.java
@@ -18,14 +18,12 @@ package com.android.phone;
 
 import android.os.Bundle;
 import android.os.PersistableBundle;
-import android.preference.CheckBoxPreference;
 import android.preference.Preference;
 import android.preference.PreferenceActivity;
 import android.preference.PreferenceScreen;
 import android.telephony.CarrierConfigManager;
 import android.view.MenuItem;
 
-import com.android.internal.telephony.Phone;
 import com.android.internal.telephony.PhoneConstants;
 
 public class GsmUmtsCallOptions extends PreferenceActivity {
@@ -64,25 +62,45 @@ public class GsmUmtsCallOptions extends PreferenceActivity {
     }
 
     public static void init(PreferenceScreen prefScreen, SubscriptionInfoHelper subInfoHelper) {
-        Preference callForwardingPref = prefScreen.findPreference(CALL_FORWARDING_KEY);
-        callForwardingPref.setIntent(subInfoHelper.getIntent(GsmUmtsCallForwardOptions.class));
-
-        Preference additionalGsmSettingsPref =
-                prefScreen.findPreference(ADDITIONAL_GSM_SETTINGS_KEY);
-        additionalGsmSettingsPref.setIntent(
-                subInfoHelper.getIntent(GsmUmtsAdditionalCallOptions.class));
-
-        Preference callBarringPref = prefScreen.findPreference(CALL_BARRING_KEY);
         PersistableBundle b = null;
         if (subInfoHelper.hasSubId()) {
             b = PhoneGlobals.getInstance().getCarrierConfigForSubId(subInfoHelper.getSubId());
         } else {
             b = PhoneGlobals.getInstance().getCarrierConfig();
         }
-        if (b != null && b.getBoolean(CarrierConfigManager.KEY_CALL_BARRING_VISIBILITY_BOOL)) {
-            callBarringPref.setIntent(subInfoHelper.getIntent(GsmUmtsCallBarringOptions.class));
-        } else {
-            prefScreen.removePreference(callBarringPref);
+
+        Preference callForwardingPref = prefScreen.findPreference(CALL_FORWARDING_KEY);
+        if (callForwardingPref != null) {
+            if (b != null && b.getBoolean(
+                    CarrierConfigManager.KEY_CALL_FORWARDING_VISIBILITY_BOOL)) {
+                callForwardingPref.setIntent(
+                        subInfoHelper.getIntent(GsmUmtsCallForwardOptions.class));
+            } else {
+                prefScreen.removePreference(callForwardingPref);
+            }
+        }
+
+        Preference additionalGsmSettingsPref =
+                prefScreen.findPreference(ADDITIONAL_GSM_SETTINGS_KEY);
+        if (additionalGsmSettingsPref != null) {
+            if (b != null && (b.getBoolean(
+                    CarrierConfigManager.KEY_ADDITIONAL_SETTINGS_CALL_WAITING_VISIBILITY_BOOL)
+                    || b.getBoolean(
+                    CarrierConfigManager.KEY_ADDITIONAL_SETTINGS_CALLER_ID_VISIBILITY_BOOL))) {
+                additionalGsmSettingsPref.setIntent(
+                        subInfoHelper.getIntent(GsmUmtsAdditionalCallOptions.class));
+            } else {
+                prefScreen.removePreference(additionalGsmSettingsPref);
+            }
+        }
+
+        Preference callBarringPref = prefScreen.findPreference(CALL_BARRING_KEY);
+        if (callBarringPref != null) {
+            if (b != null && b.getBoolean(CarrierConfigManager.KEY_CALL_BARRING_VISIBILITY_BOOL)) {
+                callBarringPref.setIntent(subInfoHelper.getIntent(GsmUmtsCallBarringOptions.class));
+            } else {
+                prefScreen.removePreference(callBarringPref);
+            }
         }
     }
 }
diff --git a/src/com/android/phone/MobileNetworkSettings.java b/src/com/android/phone/MobileNetworkSettings.java
index 691d9ffa1..6a26f76b6 100644
--- a/src/com/android/phone/MobileNetworkSettings.java
+++ b/src/com/android/phone/MobileNetworkSettings.java
@@ -788,13 +788,6 @@ public class MobileNetworkSettings extends Activity  {
             int max = mSubscriptionManager.getActiveSubscriptionInfoCountMax();
             mActiveSubInfos = new ArrayList<SubscriptionInfo>(max);
 
-            IntentFilter intentFilter = new IntentFilter(
-                    TelephonyIntents.ACTION_RADIO_TECHNOLOGY_CHANGED);
-            activity.registerReceiver(mPhoneChangeReceiver, intentFilter);
-
-            activity.getContentResolver().registerContentObserver(ENFORCE_MANAGED_URI, false,
-                    mDpcEnforcedContentObserver);
-
             Log.i(LOG_TAG, "onCreate:-");
         }
 
@@ -822,6 +815,10 @@ public class MobileNetworkSettings extends Activity  {
             @Override
             public void onReceive(Context context, Intent intent) {
                 Log.i(LOG_TAG, "onReceive:");
+                if (getActivity() == null || getContext() == null) {
+                    // Received broadcast and activity is in the process of being torn down.
+                    return;
+                }
                 // When the radio changes (ex: CDMA->GSM), refresh all options.
                 updateBody();
             }
@@ -835,6 +832,10 @@ public class MobileNetworkSettings extends Activity  {
             @Override
             public void onChange(boolean selfChange) {
                 Log.i(LOG_TAG, "DPC enforced onChange:");
+                if (getActivity() == null || getContext() == null) {
+                    // Received content change and activity is in the process of being torn down.
+                    return;
+                }
                 updateBody();
             }
         }
@@ -843,11 +844,6 @@ public class MobileNetworkSettings extends Activity  {
         public void onDestroy() {
             unbindNetworkQueryService();
             super.onDestroy();
-            if (getActivity() != null) {
-                getActivity().unregisterReceiver(mPhoneChangeReceiver);
-                getActivity().getContentResolver().unregisterContentObserver(
-                        mDpcEnforcedContentObserver);
-            }
         }
 
         @Override
@@ -884,6 +880,13 @@ public class MobileNetworkSettings extends Activity  {
 
             mSubscriptionManager.addOnSubscriptionsChangedListener(mOnSubscriptionsChangeListener);
 
+            final Context context = getActivity();
+            IntentFilter intentFilter = new IntentFilter(
+                    TelephonyIntents.ACTION_RADIO_TECHNOLOGY_CHANGED);
+            context.registerReceiver(mPhoneChangeReceiver, intentFilter);
+            context.getContentResolver().registerContentObserver(ENFORCE_MANAGED_URI, false,
+                    mDpcEnforcedContentObserver);
+
             Log.i(LOG_TAG, "onResume:-");
 
         }
@@ -1220,6 +1223,10 @@ public class MobileNetworkSettings extends Activity  {
 
             mSubscriptionManager
                     .removeOnSubscriptionsChangedListener(mOnSubscriptionsChangeListener);
+
+            final Context context = getActivity();
+            context.unregisterReceiver(mPhoneChangeReceiver);
+            context.getContentResolver().unregisterContentObserver(mDpcEnforcedContentObserver);
             if (DBG) log("onPause:-");
         }
 
diff --git a/src/com/android/phone/NetworkSelectListPreference.java b/src/com/android/phone/NetworkSelectListPreference.java
index a45c364d5..6066abe99 100644
--- a/src/com/android/phone/NetworkSelectListPreference.java
+++ b/src/com/android/phone/NetworkSelectListPreference.java
@@ -105,7 +105,8 @@ public class NetworkSelectListPreference extends ListPreference
         new AsyncTask<Void, Void, List<String>>() {
             @Override
             protected List<String> doInBackground(Void... voids) {
-                return Arrays.asList(telephonyManager.getForbiddenPlmns());
+                String[] forbiddenPlmns = telephonyManager.getForbiddenPlmns();
+                return forbiddenPlmns != null ? Arrays.asList(forbiddenPlmns) : null;
             }
 
             @Override
diff --git a/src/com/android/phone/NetworkSelectSetting.java b/src/com/android/phone/NetworkSelectSetting.java
index 71162bb0f..0655a2ee9 100644
--- a/src/com/android/phone/NetworkSelectSetting.java
+++ b/src/com/android/phone/NetworkSelectSetting.java
@@ -160,7 +160,8 @@ public class NetworkSelectSetting extends PreferenceFragment {
         new AsyncTask<Void, Void, List<String>>() {
             @Override
             protected List<String> doInBackground(Void... voids) {
-                return Arrays.asList(mTelephonyManager.getForbiddenPlmns());
+                String[] forbiddenPlmns = mTelephonyManager.getForbiddenPlmns();
+                return forbiddenPlmns != null ? Arrays.asList(forbiddenPlmns) : null;
             }
 
             @Override
diff --git a/src/com/android/phone/PhoneInterfaceManager.java b/src/com/android/phone/PhoneInterfaceManager.java
index ae3ec4eab..a327d8883 100755
--- a/src/com/android/phone/PhoneInterfaceManager.java
+++ b/src/com/android/phone/PhoneInterfaceManager.java
@@ -1716,6 +1716,10 @@ public class PhoneInterfaceManager extends ITelephony.Stub {
         // registered cell info, so return a NULL country instead.
         final long identity = Binder.clearCallingIdentity();
         try {
+            if (phoneId == SubscriptionManager.INVALID_PHONE_INDEX) {
+                // Get default phone in this case.
+                phoneId = SubscriptionManager.DEFAULT_PHONE_INDEX;
+            }
             final int subId = mSubscriptionController.getSubIdUsingPhoneId(phoneId);
             // Todo: fix this when we can get the actual cellular network info when the device
             // is on IWLAN.
@@ -2744,15 +2748,6 @@ public class PhoneInterfaceManager extends ITelephony.Stub {
         return PhoneFactory.getImsResolver().getImsConfig(slotId, feature);
     }
 
-    /**
-     * @return true if the IMS resolver is busy resolving a binding and should not be considered
-     * available, false if the IMS resolver is idle.
-     */
-    public boolean isResolvingImsBinding() {
-        enforceModifyPermission();
-        return PhoneFactory.getImsResolver().isResolvingBinding();
-    }
-
     /**
      * Sets the ImsService Package Name that Telephony will bind to.
      *
@@ -3683,7 +3678,7 @@ public class PhoneInterfaceManager extends ITelephony.Stub {
         synchronized (mLastModemActivityInfo) {
             ModemActivityInfo info = (ModemActivityInfo) sendRequest(CMD_GET_MODEM_ACTIVITY_INFO,
                     null);
-            if (info != null) {
+            if (isModemActivityInfoValid(info)) {
                 int[] mergedTxTimeMs = new int[ModemActivityInfo.TX_POWER_LEVELS];
                 for (int i = 0; i < mergedTxTimeMs.length; i++) {
                     mergedTxTimeMs[i] =
@@ -3712,6 +3707,25 @@ public class PhoneInterfaceManager extends ITelephony.Stub {
         result.send(0, bundle);
     }
 
+    // Checks that ModemActivityInfo is valid. Sleep time, Idle time, Rx time and Tx time should be
+    // less than total activity duration.
+    private boolean isModemActivityInfoValid(ModemActivityInfo info) {
+        if (info == null) {
+            return false;
+        }
+        int activityDurationMs =
+            (int) (info.getTimestamp() - mLastModemActivityInfo.getTimestamp());
+        int totalTxTimeMs = 0;
+        for (int i = 0; i < info.getTxTimeMillis().length; i++) {
+            totalTxTimeMs += info.getTxTimeMillis()[i];
+        }
+        return (info.isValid()
+            && (info.getSleepTimeMillis() <= activityDurationMs)
+            && (info.getIdleTimeMillis() <= activityDurationMs)
+            && (info.getRxTimeMillis() <= activityDurationMs)
+            && (totalTxTimeMs <= activityDurationMs));
+    }
+
     /**
      * {@hide}
      * Returns the service state information on specified subscription.
diff --git a/src/com/android/services/telephony/TelecomAccountRegistry.java b/src/com/android/services/telephony/TelecomAccountRegistry.java
index 6c8dcaa6b..642af8569 100644
--- a/src/com/android/services/telephony/TelecomAccountRegistry.java
+++ b/src/com/android/services/telephony/TelecomAccountRegistry.java
@@ -107,9 +107,31 @@ final class TelecomAccountRegistry {
         }
 
         /**
-         * Registers the specified account with Telecom as a PhoneAccountHandle.
+         * Trigger re-registration of this account.
          */
+        public void reRegisterPstnPhoneAccount() {
+            PhoneAccount newAccount = buildPstnPhoneAccount(mIsEmergency, mIsDummy);
+            if (!newAccount.equals(mAccount)) {
+                Log.i(this, "reRegisterPstnPhoneAccount: subId: " + getSubId()
+                        + " - re-register due to account change.");
+                mTelecomManager.registerPhoneAccount(newAccount);
+                mAccount = newAccount;
+            } else {
+                Log.i(this, "reRegisterPstnPhoneAccount: subId: " + getSubId() + " - no change");
+            }
+        }
+
         private PhoneAccount registerPstnPhoneAccount(boolean isEmergency, boolean isDummyAccount) {
+            PhoneAccount account = buildPstnPhoneAccount(mIsEmergency, mIsDummy);
+            // Register with Telecom and put into the account entry.
+            mTelecomManager.registerPhoneAccount(account);
+            return account;
+        }
+
+        /**
+         * Registers the specified account with Telecom as a PhoneAccountHandle.
+         */
+        private PhoneAccount buildPstnPhoneAccount(boolean isEmergency, boolean isDummyAccount) {
             String dummyPrefix = isDummyAccount ? "Dummy " : "";
 
             // Build the Phone account handle.
@@ -312,9 +334,6 @@ final class TelecomAccountRegistry {
                     .setGroupId(groupId)
                     .build();
 
-            // Register with Telecom and put into the account entry.
-            mTelecomManager.registerPhoneAccount(account);
-
             return account;
         }
 
@@ -322,6 +341,10 @@ final class TelecomAccountRegistry {
             return mAccount != null ? mAccount.getAccountHandle() : null;
         }
 
+        public int getSubId() {
+            return mPhone.getSubId();
+        }
+
         /**
          * Determines from carrier configuration whether pausing of IMS video calls is supported.
          *
@@ -579,19 +602,27 @@ final class TelecomAccountRegistry {
         }
     };
 
-    private final BroadcastReceiver mUserSwitchedReceiver = new BroadcastReceiver() {
+    private final BroadcastReceiver mReceiver = new BroadcastReceiver() {
         @Override
         public void onReceive(Context context, Intent intent) {
-            Log.i(this, "User changed, re-registering phone accounts.");
+            if (Intent.ACTION_USER_SWITCHED.equals(intent.getAction())) {
+                Log.i(this, "User changed, re-registering phone accounts.");
 
-            int userHandleId = intent.getIntExtra(Intent.EXTRA_USER_HANDLE, 0);
-            UserHandle currentUserHandle = new UserHandle(userHandleId);
-            mIsPrimaryUser = UserManager.get(mContext).getPrimaryUser().getUserHandle()
-                    .equals(currentUserHandle);
+                int userHandleId = intent.getIntExtra(Intent.EXTRA_USER_HANDLE, 0);
+                UserHandle currentUserHandle = new UserHandle(userHandleId);
+                mIsPrimaryUser = UserManager.get(mContext).getPrimaryUser().getUserHandle()
+                        .equals(currentUserHandle);
 
-            // Any time the user changes, re-register the accounts.
-            tearDownAccounts();
-            setupAccounts();
+                // Any time the user changes, re-register the accounts.
+                tearDownAccounts();
+                setupAccounts();
+            } else if (CarrierConfigManager.ACTION_CARRIER_CONFIG_CHANGED.equals(
+                    intent.getAction())) {
+                Log.i(this, "Carrier-config changed, checking for phone account updates.");
+                int subId = intent.getIntExtra(SubscriptionManager.EXTRA_SUBSCRIPTION_INDEX,
+                        SubscriptionManager.INVALID_SUBSCRIPTION_ID);
+                handleCarrierConfigChange(subId);
+            }
         }
     };
 
@@ -814,8 +845,10 @@ final class TelecomAccountRegistry {
 
         // Listen for user switches.  When the user switches, we need to ensure that if the current
         // use is not the primary user we disable video calling.
-        mContext.registerReceiver(mUserSwitchedReceiver,
-                new IntentFilter(Intent.ACTION_USER_SWITCHED));
+        IntentFilter filter = new IntentFilter();
+        filter.addAction(Intent.ACTION_USER_SWITCHED);
+        filter.addAction(CarrierConfigManager.ACTION_CARRIER_CONFIG_CHANGED);
+        mContext.registerReceiver(mReceiver, filter);
 
         // Listen to the RTT system setting so that we update it when the user flips it.
         ContentObserver rttUiSettingObserver = new ContentObserver(
@@ -951,4 +984,27 @@ final class TelecomAccountRegistry {
             mAccounts.clear();
         }
     }
+
+    /**
+     * Handles changes to the carrier configuration which may impact a phone account.  There are
+     * some extras defined in the {@link PhoneAccount} which are based on carrier config options.
+     * Only checking for carrier config changes when the subscription is configured runs the risk of
+     * missing carrier config changes which happen later.
+     * @param subId The subid the carrier config changed for, if applicable.  Will be
+     *              {@link SubscriptionManager#INVALID_SUBSCRIPTION_ID} if not specified.
+     */
+    private void handleCarrierConfigChange(int subId) {
+        if (subId == SubscriptionManager.INVALID_SUBSCRIPTION_ID) {
+            return;
+        }
+        synchronized (mAccountsLock) {
+            for (AccountEntry entry : mAccounts) {
+                if (entry.getSubId() == subId) {
+                    Log.d(this, "handleCarrierConfigChange: subId=%d, accountSubId=%d", subId,
+                            entry.getSubId());
+                    entry.reRegisterPstnPhoneAccount();
+                }
+            }
+        }
+    }
 }
diff --git a/src/com/android/services/telephony/TelephonyConnection.java b/src/com/android/services/telephony/TelephonyConnection.java
index 73bad4771..de8a52895 100644
--- a/src/com/android/services/telephony/TelephonyConnection.java
+++ b/src/com/android/services/telephony/TelephonyConnection.java
@@ -539,6 +539,7 @@ abstract class TelephonyConnection extends Connection implements Holdable {
         @Override
         public void onRttModifyResponseReceived(int status) {
             updateConnectionProperties();
+            refreshConferenceSupported();
             if (status == RttModifyStatus.SESSION_MODIFY_REQUEST_SUCCESS) {
                 sendRttInitiationSuccess();
             } else {
@@ -555,7 +556,12 @@ abstract class TelephonyConnection extends Connection implements Holdable {
 
         @Override
         public void onRttInitiated() {
-            updateConnectionProperties();
+            if (mOriginalConnection != null) {
+                // if mOriginalConnection is null, the properties will get set when
+                // mOriginalConnection gets set.
+                updateConnectionProperties();
+                refreshConferenceSupported();
+            }
             sendRttInitiationSuccess();
         }
 
@@ -564,6 +570,12 @@ abstract class TelephonyConnection extends Connection implements Holdable {
             updateConnectionProperties();
             sendRttSessionRemotelyTerminated();
         }
+
+        @Override
+        public void onOriginalConnectionReplaced(
+                com.android.internal.telephony.Connection newConnection) {
+            setOriginalConnection(newConnection);
+        }
     };
 
     protected com.android.internal.telephony.Connection mOriginalConnection;
@@ -2122,6 +2134,9 @@ abstract class TelephonyConnection extends Connection implements Holdable {
         if (mTreatAsEmergencyCall) {
             isConferenceSupported = false;
             Log.d(this, "refreshConferenceSupported = false; emergency call");
+        } else if (isRtt()) {
+            isConferenceSupported = false;
+            Log.d(this, "refreshConferenceSupported = false; rtt call");
         } else if (!isConferencingSupported || isIms && !isImsConferencingSupported) {
             isConferenceSupported = false;
             Log.d(this, "refreshConferenceSupported = false; carrier doesn't support conf.");
-- 
2.17.1

